ITR/data/base_providers.py

```
@@ -55,47 +60,61 @@ def _get_projected_production(self, scope: EScope = EScope.S1S2) -> pd.DataFrame
-        company_production = company_sector_region_info[self.column_config.BASE_YEAR_PRODUCTION]
-        return benchmark_production_projections.add(1).cumprod(axis=1).mul(
-            company_production, axis=0)
-
-    def get_benchmark_projections(self, company_sector_region_info: pd.DataFrame,
-                                  scope: EScope = EScope.S1S2) -> pd.DataFrame:
+    def get_company_projected_production(self, company_sector_region_scope: pd.DataFrame) -> pd.DataFrame:
+        """
+        get the projected productions for list of companies
+        :param company_sector_region_scope: DataFrame with at least the following columns :
+        ColumnsConfig.COMPANY_ID, ColumnsConfig.SECTOR, ColumnsConfig.REGION, ColumnsConfig.SCOPE
+        :return: DataFrame of projected productions for [base_year through 2050]
+        """
+        # get_benchmark_projections is an expensive call.  It's designed to return ALL benchmark info for ANY sector/region combo passed
+        # and it does all that work whether we need all the data or just one row.  Best to lift this out of any inner loop
+        # and use the valuable DataFrame it creates.
+        company_benchmark_projections = self.get_benchmark_projections(company_sector_region_scope)
+        company_production = company_sector_region_scope[[self.column_config.SCOPE, self.column_config.BASE_YEAR_PRODUCTION]].set_index(self.column_config.SCOPE, append=True)
+        company_production = company_production[self.column_config.BASE_YEAR_PRODUCTION]
+        if ITR.isnan(company_production.values[0].m):
+            breakpoint()
+        try:
+            # Similarly, we could pre-compute all but the final multiplication of this 30-term add/mult-add computation
+            return company_benchmark_projections.add(1).cumprod(axis=1).mul(
+                company_production, axis=0)
+        except TypeError:
+            breakpoint()
```

here, we mute the `TypeError` exception.
maybe we should let it raise?